### ISSUE TITLE
Trilinos 13.4 release status

### ISSUE THREAD

**Issue by jwillenbring** [trilinos/Trilinos#10191]
@trilinos/developers 

We are currently waiting on the 13.4 minor release until we resolve the issues associated with PRs that are not yet in develop.

**Comment by bartlettroscoe** [comment]
@jwillenbring, any update on the ETA for when the Trilinos 13.4 release branch might get cut?  I am wanting on that before I can merge PR #9978 which we know will break backward compatibility (and will straight-up break some builds and require users to take specific actions to resolve the issues).

BTW: the release notes documentation for this change in backward compatibility and how users can deal with it is described in an updated Trilinos/RELEASE_NOTES item [here](https://github.com/bartlettroscoe/Trilinos/blob/3281be9b5af148f16d9f507519d3e06b916b675c/RELEASE_NOTES#L7-L73).

If any Trilinos developers would be willing to look at that documentation and provide feedback in that PR, it would be appreciated.

**Comment by jwillenbring** [comment]
@bartlettroscoe The PRs that we were waiting on got in, but then last week there was a flurry of PRs to fix issues with things that were deprecated in the first set of PRs. We are going to reevaluate this week. There are a couple failing MueLu tests yet, but I am not yet certain if those will be fixed for the minor release.

**Comment by cgcgcg** [comment]
@jwillenbring I was only aware of one failing MueLu test. Can you point me to a listing?

**Comment by jwillenbring** [comment]
@cgcgcg I don't have a list, but I asked @csiefer2 who said that you have a PR up that might fix the issues.

**Comment by cgcgcg** [comment]
NVM then :-) I already talked to @csiefer2.

**Comment by bartlettroscoe** [comment]
Any updates on this?   We are still waiting to merge PR #9978 till after the Trilinos 13.4 release branch is created off of 'master'.  I would love to get that merged before more TriBITS changes get applied (that will also break backward compatibility in very rare cases).

**Comment by jwillenbring** [comment]
@ccober6 
@rppawlo 
@srajama1 
@mperego 
@csiefer2 
@trilinos/framework 

I went through the errors in the deprecated code off build today:

https://testing.sandia.gov/cdash/viewBuildError.php?buildid=10764652

Here are some representative failures from each package I saw issues with:

stokhos:
In file included from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/stokhos/test/UnitTest/Stokhos_TpetraCrsMatrixMPVectorUnitTest_OpenMP.cpp:46:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/stokhos/test/UnitTest/Stokhos_TpetraCrsMatrixMPVectorUnitTest.hpp: In instantiation of ‘void Tpetra_CrsMatrix_MP_VectorAdd_UnitTest<Storage, LocalOrdinal, GlobalOrdinal, Node>::runUnitTestImpl(Teuchos::FancyOStream&, bool&) const [with Storage = Stokhos::StaticFixedStorage<int, double, 16, Kokkos::OpenMP>; LocalOrdinal = int; GlobalOrdinal = long long int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP>; Teuchos::FancyOStream = Teuchos::basic_FancyOStream<char>]’:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/stokhos/test/UnitTest/Stokhos_TpetraCrsMatrixMPVectorUnitTest_OpenMP.cpp:53:1:   required from here
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/stokhos/test/UnitTest/Stokhos_TpetraCrsMatrixMPVectorUnitTest.hpp:153:48: error: ‘const class Tpetra::Map<int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >’ has no member named ‘getNodeElementList’; did you mean ‘getLocalElementList’?
   ArrayView<const GlobalOrdinal> myGIDs = map->getNodeElementList();
                                           ~~~~~^~~~~~~~~~~~~~~~~~
                                           getLocalElementList

Moertel (may not need to fix this):

In file included from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/moertelmortar/Moertel_ManagerT.cpp:52:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/moertel/src/mortar/Moertel_ManagerT_Def.hpp: In instantiation of ‘bool MoertelT::ManagerT<ST, LO, GO, N>::BuildSaddleMap() [with ST = double; LO = int; GO = long long int; N = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>]’:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/moertelmortar/Moertel_ManagerT.cpp:56:3:   required from here
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/moertelmortar/Moertel_ManagerT_Def.hpp:1656:43: error: ‘const class Tpetra::Map<int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >’ has no member named ‘getNodeNumElements’; did you mean ‘getLocalNumElements’?
   size_t nummyelements     = problemmap_->getNodeNumElements() +
                              ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
                              getLocalNumElements

Shylu:
In file included from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/shylu/shylu_dd/bddcshylu_UtilBDDC.hpp:59,
                 from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/shylu/shylu_dd/bddcshylu_PartitionOfUnityBDDC.hpp:52,
                 from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/shylu/shylu_dd/bddcshylu_PreconditionerBDDC.hpp:54,
                 from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/shylu/shylu_dd/bddc/test/SolverBDDCTestSimple.cpp:48:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/shylu/shylu_dd/bddc/src/shylu_PComm.hpp: In member function ‘void bddc::PComm<LO, GO, SX>::checkSizes()’:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/shylu/shylu_dd/bddcshylu_PComm.hpp:368:45: error: no matching function for call to ‘Tpetra::Distributor::doPostsAndWaits(Teuchos::ArrayView<const long unsigned int>, int, Teuchos::ArrayView<long unsigned int>)’
        Teuchos::ArrayView<size_t>(recvCount));
                                             ^

ROL (not sure why the output seems to be missing characters here):

In file included from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packagesexample/PDE-OPT/0ld/elasticity/../TOOLS/elasticity.hpp:50,
                 from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packagesexample/PDE-OPT/0ld/elasticity/data.hpp:49,
                 from /scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packagesexample/PDE-OPT/0ld/elasticity/example_01.cpp:65:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packages/rol/example/PDE-OPT/0ld/elasticity/../TOOLS/../TOOLS/PDE_FEM.hpp: In instantiation of ‘void PDE_FEM<Real>::EnforceDBC() [with Real = double]’:
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packagesexample/PDE-OPT/0ld/elasticity/data.hpp:77:2:   required from ‘ElasticityData<Real>::ElasticityData(ROL::Ptr<const Teuchos::Comm<int> >&, const Teuchos::RCP<Teuchos::ParameterList>&, ROL::Ptr<std::basic_ostream<char> >&) [with Real = double; ROL::Ptr<const Teuchos::Comm<int> > = Teuchos::RCP<const Teuchos::Comm<int> >; ROL::Ptr<std::basic_ostream<char> > = Teuchos::RCP<std::basic_ostream<char> >]’
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packagessrc/compatibility/teuchos/rcp/ROL_Ptr.hpp:83:24:   required from ‘ROL::Ptr<T> ROL::makePtr(Args&& ...) [with T = ElasticityData<double>; Args = {Teuchos::RCP<const Teuchos::Comm<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::RCP<std::basic_ostream<char, std::char_traits<char> > >&}; ROL::Ptr<T> = Teuchos::RCP<ElasticityData<double> >]’
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packagesexample/PDE-OPT/0ld/elasticity/example_01.cpp:98:105:   required from here
/scratch/trilinos/jenkins/ascic113/workspace/trilinos-folder/gcc-8.3.0_nightly_deprecated_off/Trilinos/packagesexample/PDE-OPT/0ld/elasticity/../TOOLS/../TOOLS/PDE_FEM.hpp:699:11: error: no matching function for call to ‘Tpetra::CrsMatrix<>::getGlobalRowCopy(PDE_FEM<double>::GO&, Teuchos::Array<long long int>&, Teuchos::Array<double>&, size_t&)’
           matA_dirichlet_->getGlobalRowCopy(gDof, indices, values, numRowEntries);
           ^~~~~~~~~~~~~~~

The total number of errors is huge. I am wondering if someone already used a script to update other packages that could be used to update these as well? Pehaps the errors emanate primarily from a small number of calls.

**Comment by cgcgcg** [comment]
I used something like this:
```
grep -r -l "getNodeNumElements" * | xargs sed -i 's/getNodeNumElements/getLocalNumElements/g'                                                                                                                                                                                                                                                         
grep -r -l "getNodeNumRows" * | xargs sed -i 's/getNodeNumRows/getLocalNumRows/g'                                                                                                                                                                                                                                                                     
grep -r -l "getNodeNumCols" * | xargs sed -i 's/getNodeNumCols/getLocalNumCols/g'                                                                                                                                                                                                                                                                     
grep -r -l "getNodeMaxNumRowEntries" * | xargs sed -i 's/getNodeMaxNumRowEntries/getLocalMaxNumRowEntries/g'                                                                                                                                                                                                                                          
grep -r -l "getNodeElementList" * | xargs sed -i 's/getNodeElementList/getLocalElementList/g'                                                                                                                                                                                                                                                         
grep -r -l "getNodeNumEntries" * | xargs sed -i 's/getNodeNumEntries/getLocalNumEntries/g'
```

**Comment by mperego** [comment]
Thanks @cgcgcg . That won't fix the issues in ROL and @trilinos/shylu.

I just realized that when I fixed ROL, I had Intrepid off, so some of the tests were not compiled.
@dridzal what's the plan with the PDE-OPT/Old tests? Can we disable them, or they are still important?

**Comment by cgcgcg** [comment]
For ShyLU: do these need to be fixed? When I configure Trilinos, I see this message:
```
CMake Warning at packages/shylu/shylu_dd/bddc/CMakeLists.txt:4 (MESSAGE):
  ***** WARNING: Package BDDC is deprecated in February 2022; it will be
  removed by next release *****
```

**Comment by srajama1** [comment]
@iyamazaki : The BDDC error needs to be fixed, right?

**Comment by dridzal** [comment]
@mperego : The `PDE-OPT/0ld` examples can be disabled at this time.  Is this limited to `PDE-OPT/0ld`, or do all examples that use finite element assembly through Tpetra need to be updated?

**Comment by mperego** [comment]
> @mperego : The `PDE-OPT/0ld` examples can be disabled at this time. Is this limited to `PDE-OPT/0ld`, or do all examples that use finite element assembly through Tpetra need to be updated?

Unfortunatey there are other examples that call Tpetra deprecated code. I'm creating a PR with a fix.

**Comment by iyamazaki** [comment]
@srajama1. I had compiled BDDC with deprecated-code off before, so I will see if these code changes can fix these errors.

**Comment by iyamazaki** [comment]
@jwillenbring. About ShyLUDD/BDDC errors, as @cgcgcg also mentioned, BDDC is going to be deprecated for the next release, and instead of fixing the issue, it will be good if we could just remove it from the test.. @srajama1

**Comment by srajama1** [comment]
I agree with @iyamazaki !

**Comment by jwillenbring** [comment]
> @jwillenbring. About ShyLUDD/BDDC errors, as @cgcgcg also mentioned, BDDC is going to be deprecated for the next release, and instead of fixing the issue, it will be good if we could just remove it from the test.. @srajama1

Is there an option we can add to the configuration to remove BDDC from this build?

**Comment by iyamazaki** [comment]
@jwillenbring Can you tell me where I can find the build script? I think BDDC is the default for ShyLU_DD. So, I think you just need to remove Trilinos_ENABLE_ShyLU_DD:BOOL=ON, and enable each solver you want to test in ShyLU_DD (which I assume you already do)? @srajama1

**Comment by e10harvey** [comment]
> @iyamazaki, I believe the deprecated off build uses GenConfig.

Correct. In the near future, all PR and mastermerge configurations will use GenConfig.

**Comment by e10harvey** [comment]
> @e10harvey might have more insight into the inner workings of that.

@ZUUL42: [Here is the configuration](https://gitlab-ex.sandia.gov/trilinos-devops-consolidation/code/GenConfig/-/blob/master/ini_files/trilinos/config-specs.ini#L1828).

**Comment by srajama1** [comment]
@iyamazaki is right, Can we tempoararily do 

Trilinos_ENABLE_ShyLU_DD_BDDC:BOOL=OFF ? 

Just this might do the trick.

**Comment by jwillenbring** [comment]
@ZUUL42 

Could you please add 

-DTrilinos_ENABLE_Moertel:BOOL=OFF
-DTrilinos_ENABLE_ShyLU_DD_BDDC:BOOL=OFF

to the deprecated code build?

**Comment by ZUUL42** [comment]
@jwillenbring GenConfig update submitted.

**Comment by jwillenbring** [comment]
@iyamazaki @srajama1 

CC: @ZUUL42 @ccober6 

It appears Trilinos_ENABLE_ShyLU_DD_BDDC:BOOL=OFF may not be a valid option. This did not have the desired impact, and I can't find this option in other contexts. Is there another option or set of options we can use to disable the code that we don't want to include in this deprecated code off build?

I can report that Moertel was successfully disabled.

**Comment by jhux2** [comment]
@jwillenbring Try`Trilinos_ENABLE_ShyLU_DDBDDC:BOOL=OFF`.  That shows as a valid option in CMakeCache.txt in one of my builds.

**Comment by bartlettroscoe** [comment]
> It appears Trilinos_ENABLE_ShyLU_DD_BDDC:BOOL=OFF may not be a valid option

NOTE: If you set a CMake cache var that is not read by CMake or TirBITS, you will get a warning at the end like shown at:

* https://testing.sandia.gov/cdash/build/10936593/configure

showing:

```
Finished configuring Trilinos!

-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    BoostLib_INCLUDE_DIRS
    BoostLib_LIBRARY_DIRS
    Boost_INCLUDE_DIRS
    Boost_LIBRARY_DIRS
    Netcdf_INCLUDE_DIRS
    Netcdf_LIBRARY_DIRS


-- Build files have been written to: /data/jhu/trilinos-test/SERIAL-DEBUG_NO-EPETRA/BUILD
```

**Comment by iyamazaki** [comment]
@jwillenbring. I did not mean `Trilinos_ENABLE_ShyLU_DDBDDC:BOOL=OFF`. 

I was wondering if you could remove `Trilinos_ENABLE_ShyLU_DD:BOOL=ON`, if you have it (ShyLU_DDBDDC is currently the default for ShuLU_DD). Then, we could enable an individual package under ShyLU_DD that we really want to test, e.g., `Trilinos_ENABLE_ShyLU_DDFROSch:BOOL=ON` (without having `Trilinos_ENABLE_ShyLU_DD:BOOL=ON`).

**Comment by mperego** [comment]
@iyamazaki I think it's better to disable the least number of modules possible. My understanding it that you prefer not to fix the DBBC code because it is going to be deprecated. That's fine, but I think it's best to disable only that module. Unless you have strong reasons to disable other modules.

**Comment by jwillenbring** [comment]
@iyamazaki What you describe may work in the very short term, but we would like to apply this also to the PR testing. In that case, the system decides what to enable based on what has changed, so there is no option to remove a particular enable, our option is to specify things to turn off. So we need options of the general type something=OFF, rather than adding anything that says something=ON, or revoving an option that has something=ON. Can we express what you want by turning things off?

**Comment by jwillenbring** [comment]
To make this a little clearer, if we did add something=ON, "something" would be built for all PR builds, regardless of whether or not it was changed by the PR being tested. There is no option to remove something=ON, because those are all specified by the system that figures out what is changed.

**Comment by iyamazaki** [comment]
Hi, @jwillenbring. @srajama1 may have a better suggestion. I thought it may be an okay 'temporary' solution since I think ShyLU_DD has just three packages (including BDDC). 

Another option may be to switch the default solver of ShyLU_DD, e.g., from BDDC to FROSch (we need to do this anyway).

**Comment by bartlettroscoe** [comment]
> To make this a little clearer, if we did add something=ON, "something" would be built for all PR builds, regardless of whether or not it was changed by the PR being tested

Right, you can only disable things.  If it is not already supported, we can make it so that explicit disables override enables based on changed (sub)packages.

A more clear what to do this might be to make make the subpackage EX.  See:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#se-package-test-group

and:

* https://tribitspub.github.io/TriBITS/maintainers_guide/index.html#ex-se-packages-disabled-by-default

I believe the line of code you would change in this case is:

https://github.com/trilinos/Trilinos/blob/develop/packages/shylu/shylu_dd/cmake/Dependencies.cmake#L5

from:

```
  BDDC                  bddc              PT       REQUIRED
```

to

```
  BDDC                  bddc              EX       OPTIONAL
```

And anything downstream for ShyLU_DDBDDC would get disabled as well.

**Comment by jwillenbring** [comment]
> A more clear what to do this might be to make make the subpackage EX. 

I think this is potentially a good idea. I suggested this for Moertel actually a few weeks ago. I felt that was even more of a slam-dunk case because of the lack of use of Moertel. My suggestion was not well-received.

**Comment by mperego** [comment]
B.t.w. The fact that BDDC is a required package means that we cannot disable it without disabling ShyLU_DD, right? That's a problem because we want to test FROSch that is part of ShyLU_DD. @srajama1 @iyamazaki can we make BDDC optional?

**Comment by bartlettroscoe** [comment]
> I think this is potentially a good idea. I suggested this for Moertel actually a few weeks ago. I felt that was even more of a slam-dunk case because of the lack of use of Moertel. My suggestion was not well-received.

It does not serve anyone to disable a package in all automated testing but still keep it and release it like it is a supported package.

**Comment by jwillenbring** [comment]
> It does not serve anyone to disable a package in all automated testing but still keep it and release it like it is a supported package.

The context was only for testing with deprecated code turned off, not for all automated testing.

**Comment by iyamazaki** [comment]
Hi, @mperego. I can turn FROSch on without ShyLU_DD on.

**Comment by mperego** [comment]
@bartlettroscoe Thanks for the explanation. If I understand correctly, in order to turn off _only_ DBBC, without  explicitly adding something ON, DBBC needs to be made OPTIONAL first, right? Then we can turn it off with `Trilinos_ENABLE_ShyLU_DDBDDC:BOOL=OFF`. Is this correct?

**Comment by srajama1** [comment]
I just want to add one additional context. BDDC is already deprecated. It will be removed after this release. It is not the best use of our time that the deprecated BDDC will hold up the release. Please do whatever is prudent to get the release out. There are no users for this subpackage and we will remove it after the release.

**Comment by jwillenbring** [comment]
>  just want to add one additional context. BDDC is already deprecated. It will be removed after this release. It is not the best use of our time that the deprecated BDDC will hold up the release. Please do whatever is prudent to get the release out. There are no users for this subpackage and we will remove it after the release.

Thanks, @srajama1. I actually just had a conversation with @ccober6 that is very consistent with this and I asked @ZUUL42 to move forward with trying Trilinos_ENABLE_ShyLU_DDBDDC:BOOL=OFF

**Comment by mperego** [comment]
@jwillenbring If I understand correctly that will turn off `ShyLU_DD`, because `BDDC` is a required package. We first need to make `BDDC` optional. 

> If you disable a required subpackage, that will disable its parent package and all downstream packages that have a required dependency on that parent package.

**Comment by bartlettroscoe** [comment]
> If I understand correctly, in order to turn off only DBBC, without explicitly adding something ON, DBBC needs to be made OPTIONAL first, right? Then we can turn it off with `Trilinos_ENABLE_ShyLU_DDBDDC:BOOL=OFF`. Is this correct?

The impact of setting  `Trilinos_ENABLE_ShyLU_DDBDDC:BOOL=OFF` depends on how the dependencies are defined in downstream packages and what you want to allow to be enabled.    And it looks like there is nothing in Trilinos that depends on `ShyLU_DDBDDC` as **not** shown by:

```
$ find packages/ -name Dependencies.cmake -exec grep -nH "ShyLU_DDBDDC" {} \;
[emtpy]
```

In fact, there does not look to be any downstream packages that depend on any of the Shy_DD subpackages as **not** shown by:

```
$ find packages/ -name Dependencies.cmake -exec grep -nH "ShyLU_DD[a-zA-Z0-9_]" {} \;
[empty]
``` 

But there are downstream dependencies for subpackages of `ShyLU_Node` as shown by:

```
$ find packages/ -name Dependencies.cmake -exec grep -nH "ShyLU_Node[a-zA-Z0-9_]" {} \;
packages/amesos2/cmake/Dependencies.cmake:6:SET(LIB_OPTIONAL_DEP_PACKAGES Epetra EpetraExt ShyLU_NodeBasker ShyLU_NodeTacho)
packages/amesos2/cmake/Dependencies.cmake:8:SET(TEST_OPTIONAL_DEP_PACKAGES ShyLU_NodeBasker ShyLU_NodeTacho Kokkos TrilinosSS)
packages/shylu/shylu_dd/bddc/cmake/Dependencies.cmake:2:SET(LIB_OPTIONAL_DEP_PACKAGES Amesos2 MueLu ShyLU_NodeTacho)
packages/ifpack2/cmake/Dependencies.cmake:2:SET(LIB_OPTIONAL_DEP_PACKAGES Xpetra Zoltan2Core ThyraTpetraAdapters Amesos2 ShyLU_NodeBasker ShyLU_NodeHTS ShyLU_NodeFastILU)
packages/ifpack2/cmake/Dependencies.cmake:4:SET(TEST_OPTIONAL_DEP_PACKAGES Amesos2 ShyLU_NodeHTS ML AztecOO Epetra)
```

The only dependency of `ShyLU_DD` is the pseudo package `ShyLU` as shown by:

```
$ find packages/ -name Dependencies.cmake -exec grep -nH "ShyLU_DD[^a-zA-Z0-0_]" {} \;
packages/shylu/cmake/Dependencies.cmake:1:SET(LIB_REQUIRED_DEP_PACKAGES ShyLU_DD ShyLU_Node)
```

And nothing depends on the pseudo package `ShyLU` as **not** shown by:

```
$ find packages/ -name Dependencies.cmake -exec grep -nH "ShyLU[^a-zA-Z0-0_]" {} \;
packages/shylu/shylu_node/cmake/Dependencies.cmake:4:  # New ShyLU subpackages:
packages/shylu/shylu_dd/cmake/Dependencies.cmake:4:  # New ShyLU subpackages:
```

Therefore, setting  `Trilinos_ENABLE_ShyLU_DDBDDC:BOOL=OFF` should only have the impact of disabling `ShyLU_DDBDDC` and its tests and disabling the downstream shell parent package `ShyLU_DD` and the pseudo package `ShyLU`.  Therefore, you should **not** have to make  `ShyLU_DDBDDC` an optional subpackage of its parent package `ShyLU_DD`.

> I just want to add one additional context. BDDC is already deprecated.

However, since `ShyLU_DDBDDC` is deprecated, it should be made optional in `package/shylu/shylu_dd/cmake/Dependencies.cmake` but it really does not matter for things inside of Trilinos.  However, it would matter for external customers because if they set `Trilinos_ENABLE_ShyLU_DDBDDC=OFF` but then try to set `Trilinos_ENABLE_ShyLU_DD=ON` or `Trilinos_ENABLE_ShyLU=ON`, then they will get a configure-time error message noting the conflict because Trilinos, by default, errors out on enable/disable conflicts due to it setting the default [`Trilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=OFF`](https://docs.trilinos.org/files/TrilinosBuildReference.html#disable-a-package-and-all-its-dependencies).

**Comment by mperego** [comment]
@bartlettroscoe Oh, OK. I thought that `FROSch` would be disabled because it's a downstream package of `ShyLU_DD`, but it won't, because `FROSch` does not depend on `ShyLU_DD`. Kind of tricky but I understand why we want that behavior. Thanks again for the explanation.

**Comment by bartlettroscoe** [comment]
> Oh, OK. I thought that FROSch would be disabled because it's a downstream package of ShyLU_DD, but it won't, because FROSch does not depend on ShyLU_DD

Right, it is the other way around, ShyLU_DD depends on ShyLU_DDFROSch so that if someone sets a dependency on ShyLU_DD, they will get ShyLU_DDFROSch.  Trying to make the enable/disable behavior and relationship of package and subpackages seem logical to everyone has been challenging.

**Comment by jwillenbring** [comment]
A branch has been created for the 13.4 release. Next steps are to notify users that backward compatibility will be dropped soon on develop, and then communicate to Trilinos devleopers a date when this can begin happening. As for the release, we also need to update release versions on the branch (and on develop) and ultimately tag an official 13.4 release version.

**Comment by jwillenbring** [comment]
Trilinos 13.4 has been tagged. The release has been completed.
